CIP-0141? | Web-Wallet Bridge - Plutus wallets

leo42 opened the pull request:
This a specification for a CIP-30 extension for plutus based wallets. 

---

([current rendered proposal](https://github.com/leo42/CIPs/blob/Pulutus-Wallet-connector/CIP-0141/README.md))

rphair commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,218 @@
+---
+CIP: ???
+Title: Web-Wallet Bridge - Plutus based wallets
```
```suggestion
Title: Web-Wallet Bridge - Plutus wallets
```

rphair replied:
resolving: updated as suggested.

rphair commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,218 @@
+
+## Abstract
+
+This document describes a CIP-30 extension allowing webpages (i.e. dApps) to interface with Cardano Plutus-wallets. This document is a work in progress and is not yet finalized. It is expected to be updated as the ecosystem evolves. 
```
```suggestion
This document describes a CIP-30 extension allowing webpages (i.e. dApps) to interface with Cardano Plutus wallets. This interface is a work in progress and new versions are expected to be included as the ecosystem evolves. 
```
beyond the "Plutus wallets" suggestion... the original text gives us no grounds for merging or activating this CIP.  Does this revision reflect what you state below in the Versioning section?

rphair commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,218 @@
+
+## Motivation: why is this CIP necessary?
+
+In order to facilitate future dApp development, we will need a way for dApps to communicate with plutus wallets, given the unique complexities of plutus based addresses. Special provisions need to be made to make the connector compatible with them.  
```
```suggestion
In order to facilitate future dApp development, we will need a way for dApps to communicate with Plutus wallets, given the unique complexities of Plutus based addresses. Special provisions need to be made to make the connector compatible with them.  
```

rphair replied:
resolving: updated as suggested.

rphair commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,218 @@
+
+Additionally, apps need to be able to submit a transaction to the wallet for signing in an asynchronous manner, as gathering of signatures can take a long time and each wallet provider will have its own way of handling this process. 
+
+Finally, the signData() endpoint will have to be disabled when using this extension since they are not compatible with plutus based addresses. plutus contracts cannot sign data, only validate transactions.
```
```suggestion
Finally, the `signData()` endpoint will have to be disabled when using this extension since they are not compatible with plutus based addresses. plutus contracts cannot sign data, only validate transactions.
```
... and throughout, whenever functions or parameters are quoted?

rphair replied:
resolving: updated as suggested.

rphair commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,218 @@
+### Rationale for the required data
+
+- Script descriptor:
+    Any transaction consuming a UTxO from a plutus based address must attach the corresponding script. 
```
```suggestion
    Any transaction consuming a UTxO from a Plutus based address must attach the corresponding script. 
```
... etc, throughout the document.

rphair replied:
resolving: updated as suggested.

rphair commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,218 @@
+
+Finally, the `signTx()` and `signData()` endpoints will have to be disabled when using this extension since they are not compatible with native-script-based addresses.
+
+## Rationale for the required data
```
@leo42 I think I see why you've written it in this order (to keep the rationale close to the data structure it refers to) but I still think you should move it down into the usual _Rationale_ section which currently contains nothing but a reference back here.

Remember also you state that this document is a "work in progress" which means this rationale might not always apply, nor apply to future versions.  Therefore as each section grows it will help us (editors, implementors, and future version authors) to maintain the standard document organisation without forward & backward references.

rphair replied:
resolving: moved to proper _Rationale_ section as suggested.

rphair commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,218 @@
+
+## Rationale: how does this CIP achieve its goals?
+
+See justification and explanations provided with each API endpoint.
```
see comment to be anchored by GitHub later

rphair reviewed (commented):
@leo42 this looks very useful. 🙏 Can we make the terminology consistent between the PR, header, and document text?  "Plutus wallets" seems functionally identical to "Plutus based wallets" and is more concise.

leo42 commented:
Thank you for the feedback @rphair , I updated the document with all the suggestions and passed it though a spellcheck 1 more time.

rphair commented:
thanks @leo42 - I've put this for introduction at next CIP Editors Meeting (https://hackmd.io/@cip-editors/87) & in the meantime the technical review can continue here.

rphair commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,216 @@
+
+Specifically, apps building transactions need to be able to get the following information from the wallet:
+
+- Script descriptor
+- `ScriptRequirements`
+- Change Datum
+
+Additionally, apps need to be able to submit a transaction to the wallet for signing in an asynchronous manner, as gathering of signatures can take a long time and each wallet provider will have its own way of handling this process. 
+
+Finally, the `signData()` endpoint will have to be disabled when using this extension since they are not compatible with Plutus-based addresses. Plutus contracts cannot sign data, only validate transactions.
+
+### Rationale for the required data
+
+- Script descriptor:
+    Any transaction consuming a UTXO from a Plutus-based address must attach the corresponding script. 
+
+- `ScriptRequirements`:
+    The `TxContext` that is required to be able to validate the transaction. It encompasses all the possible combinations of requirements for the transaction to be valid, as such it is represented by an array of `ScriptRequirement` objects.
+
+- Change Datum:
+    The datum that will be used as the change output for the transaction. This is required for wallets based on Plutus V2 and before, as the change output must contain a datum to be valid and spendable.
```
```suggestion
- Script descriptor
  - Any transaction consuming a UTXO from a Plutus-based address must attach the corresponding script. 
- `ScriptRequirements`
  - The `TxContext` that is required to be able to validate the transaction. It encompasses all the possible combinations of requirements for the transaction to be valid, as such it is represented by an array of `ScriptRequirement` objects.
- Change Datum
  - The datum that will be used as the change output for the transaction. This is required for wallets based on Plutus V2 and before, as the change output must contain a datum to be valid and spendable.

Additionally, apps need to be able to submit a transaction to the wallet for signing in an asynchronous manner, as gathering of signatures can take a long time and each wallet provider will have its own way of handling this process. 

Finally, the `signData()` endpoint will have to be disabled when using this extension since they are not compatible with Plutus-based addresses. Plutus contracts cannot sign data, only validate transactions.
```
Those present at yesterday's meeting got a bit dizzy thinking there were 2 _Rationale_ document sections & flipping back & forth.  These 3 items below aren't "rationale" as we've defined it in the CIP process (comparative reasons why design decisions were taken).

I think this problem could be fixed & would produce a more readable CIP if each of these 3 items were moved up with the corresponding item above.  I'm taking a shot at correcting this & I'll commit it to confirm it's readable & feel free to make further changes afterward.

rphair reviewed (commented):
@leo42 - in the CIP meeting yesterday there was the big concern that changing the CIP-0030 API won't just provide additional functionality for the Plutus wallets... it will require modifying all non-Plutus wallets as well, since the API would also be changing for _them_.

If this is true then it seems we would have a stumbling block with writing this as a CIP-0030 extension and therefore would be a more readily implemented if it were a standalone CIP.

@Crypto2099 @Ryun1 I hope that captures the spirit of the problem & please feel free to elaborate or correct.

leo42 commented:
> @leo42 - in the CIP meeting yesterday there was the big concern that changing the CIP-0030 API won't just provide additional functionality for the Plutus wallets... it will require modifying all non-Plutus wallets as well, since the API would also be changing for _them_.
> 
> If this is true then it seems we would have a stumbling block with writing this as a CIP-0030 extension and therefore would be a more readily implemented if it were a standalone CIP.
> 
> @Crypto2099 @Ryun1 I hope that captures the spirit of the problem & please feel free to elaborate or correct.

Hey @rphair, sorry I could not make the CIP editors meeting I will try to be there for the next one.

The design follows the structure of [CIP-106,](https://github.com/cardano-foundation/CIPs/pull/617) offering a modified API set If and only If this CIP is enabled, this way both the app developers and wallet developers have the freedom to support it or not.

If the CIP number(not yet assigned) is not included in the connection request then the API is expected to conform to the CIP-30 standard or reject the connection.

rphair commented:
> @leo42 https://github.com/cardano-foundation/CIPs/pull/798#issuecomment-2088439304: The design follows the structure of https://github.com/cardano-foundation/CIPs/pull/617 offering a modified API set If and only If this CIP is enabled, this way both the app developers and wallet developers have the freedom to support it or not.

@Ryun1 @Crypto2099 this is also what I was thinking at the meeting... doesn't in fact CIP-0030 require presenting a set of supported CIPs at the time a wallet opens an API connection?  And if so, wouldn't that avoid any conflicts with conventional wallets... which of course would _not_ indicate support for this CIP as they would if they were Plutus wallets?

Hence let me say again I'm not the best one to identify the potential problem & perhaps you guys can take it from here 😇

Ryun1 commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,208 @@
+* NotFound - The transaction with the given id was not found.
+* NotReady - The transaction with the given id is not ready yet. 
+
+## Motivation: why is this CIP necessary?
```
Duplicate section

rphair replied:
Not really a duplicate section but fixed in https://github.com/cardano-foundation/CIPs/pull/798/commits/4805b23d537f3a8c3e35fdb7731ab27e53a5138a.

Ryun1 commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,208 @@
+
+### V2 Additional API Endpoints
+
+#### `api.getScriptRequirements`: Promise<ScriptRequirement[]>
```
```suggestion
#### `api.cipxxxx.getScriptRequirements`: Promise<ScriptRequirement[]>
```
endpoints should be name spaced, to follow CIP30

Ryun1 commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,208 @@
+
+For wallets with multiple spend conditions, separate entries in the list should be used to represent each spend condition. Wallet providers should implement UX to allow users to order the list of `ScriptRequirement` from most to least preferred. DApps should use the first entry in the list that is valid for the current transaction or select one based on the logic of their use-case.
+
+#### `api.getScript()`: Promise[<number<plutusVersion>],<CBOR<plutusScript>>
```
```suggestion
#### `api.getScript()`: Promise[<number<plutusVersion>>],<CBOR<plutusScript>>
```

rjharmon commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,208 @@
+
+## Motivation: why is this CIP necessary?
+
+In order to facilitate future DApp development, we will need a way for DApps to communicate with Plutus wallets, given the unique complexities of Plutus-based addresses. Special provisions need to be made to make the connector compatible with them.  
```
needs more explicit info about what "Plutus wallet" actually means.

leo42 replied:
Added a section explaining Plutus wallets and a few possible implementation of them

rphair commented:
@leo42 it looks like your last round of commits might address all the pending issues with this proposal & therefore should be considered as a potential candidate again.

If it is indeed ready for re-review, we would have noticed this sooner if you'd posted a comment (something like "I'm done with my commits, here's the purpose of the changes, please let's continue review")... though we are working on shrinking the PR queue now so the pending PRs will stand out more in the future.

I'm marking this `Unconfirmed` and will therefore put it onto the CIP meeting agenda if you believe it's ready for review again.  Please in any case summarise the last changes you've made, so we can expedite re-reviewing when that time comes.

leo42 commented:
> @leo42 it looks like your last round of commits might address all the pending issues with this proposal & therefore should be considered as a potential candidate again.
> 
> If it is indeed ready for re-review, we would have noticed this sooner if you'd posted a comment (something like "I'm done with my commits, here's the purpose of the changes, please let's continue review")... though we are working on shrinking the PR queue now so the pending PRs will stand out more in the future.
> 
> I'm marking this `Unconfirmed` and will therefore put it onto the CIP meeting agenda if you believe it's ready for review again. Please in any case summarise the last changes you've made, so we can expedite re-reviewing when that time comes.

@rphair It is indeed ready for a re-review. 

In the latest changes I added support for wallets that use external keys + secrets encrypted in the datum to control a set of UTxOs, in this case the data should be opaque until the User signs of on allowing the dap to spend the UTxO. 

I am on hold for the hardfork, after which I will implement a reference wallet and its connector where I might stubble into problems that are not currently obvious. 

I also solicited feedback from other teams planning smart wallets but nothing is coming through atm.

rphair commented:
@leo42 could you please respond to @Ryun1's 2 points about namespacing: https://github.com/cardano-foundation/CIPs/pull/798#discussion_r1600347350 & https://github.com/cardano-foundation/CIPs/pull/798#discussion_r1600352296

... and then I would feel comfortable passing this myself & getting this marked as `Last Check` for the next CIP meeting.  Apologies for not being able to progress this when you were at the meeting today but we did have about 100 things come up in the last couple days when dredging the PR queue. :)

rphair commented:
Waiting for author to resume work on this (tagging as such).

leo42 commented:
I just updated with the final version, made some last minute changes while creating the reference implementation that I did not think off before.

I also added support for externally owned address( address controlled by keys not native to Cardano) 

Tagging @elRaulito @MicroProofs working on relevant products to confirm this can/will support your needs

rphair commented:
thanks @leo42, those changes look significant: I've put it on Review for the next CIP meeting (https://hackmd.io/@cip-editors/101) so we can see about assigning this a CIP number.

Ryun1 commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,219 @@
+Category: Wallets
+Authors: 
+  - Leo
+Implementors: NA
```
```suggestion
Implementors: BroClan
```

rphair replied:
Resolved with https://github.com/cardano-foundation/CIPs/pull/798/commits/404dc6351e8fff3ff316e4f247d6d1d166f29eb6.

Ryun1 commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,219 @@
+Authors: 
+  - Leo
+Implementors: NA
+Discussions:
```
```suggestion
Discussions:
  - https://github.com/cardano-foundation/CIPs/pull/798
```

rphair commented on CIP-xxxxx/README.md:
```diff
@@ -0,0 +1,220 @@
+---
+CIP: ???
```
```suggestion
CIP: 141
```

rphair reviewed (commented):
Those present at the CIP meeting today were happy to see this moving along & hope its declared candidacy will invite discussions from dApp developers about how they intend to interact with this proposed standard... so please @leo42 invite any interested parties for review.

If there is some interest on the CIP Discord we might also create a channel for it, as we have for other CIP specific online working groups so far (@Ryun1 at least will keep an eye out for this).

Please rename your containing folder to `CIP-0141` and update the above link to your README file (in the original comment) accordingly. 🎉

rphair reviewed (approved):
@leo42 this got stalled many moons ago because the editors were waiting, I think as a bare minimum, to add your own team as an implementor as @Ryun1  suggested in https://github.com/cardano-foundation/CIPs/pull/798#discussion_r1858848296 ... this would have made it look "abandoned" but I can see from GitHub this is actively maintained & live in the Cardano community.

So I think the editors need to put this on the boards even though so far you are the only implementor.  I believe it has potential as a "community" standard even with this one implementation, although please keep in mind to be `Active` the current precedent would require adoption elsewhere as well.

The document itself needed a general format cleanup (https://github.com/cardano-foundation/CIPs/pull/798/commits/404dc6351e8fff3ff316e4f247d6d1d166f29eb6) (including a broken link fix to your own repository) so I'm submitting this in my own commit & suggest we move this forward to merge via `Last Check` status at the next CIP meeting in a couple days (https://hackmd.io/@cip-editors/113) so please keep that mind if there are any last-minute updates necessary.

Ryun1 commented on CIP-0141/README.md:
```diff
@@ -0,0 +1,216 @@
+
+Returns the CBOR-encoded Plutus script that controls this wallet.
+
+#### `api.cipxxxx.submitUnsignedTx(tx: CBOR<unsignedTransaction>)`: Promise<hash32>
```
```suggestion
#### `api.cip141.submitUnsignedTx(tx: CBOR<unsignedTransaction>)`: Promise<hash32>
```

Ryun1 commented on CIP-0141/README.md:
```diff
@@ -0,0 +1,216 @@
+
+If the transaction is not ready, the wallet should throw a `CompletedTxError` with the appropriate error code. If the transaction is ready, the wallet should return the CBOR-encoded transaction and the signatures.
+
+#### `api.cipxxxx.getSecret(secretId: String)`: Promise<String>
```
```suggestion
#### `api.cip141.getSecret(secretId: String)`: Promise<String>
```

Ryun1 commented on CIP-0141/README.md:
```diff
@@ -0,0 +1,216 @@
+
+Wallets should request authorization from the user before returning the secret to the DApp.
+
+#### `api.cipxxxx.signRedeemer(data: Data, primitive : encriptionPrimitive  )`: Promise<CBOR<redeemer>>
```
```suggestion
#### `api.cip141.signRedeemer(data: Data, primitive : encriptionPrimitive  )`: Promise<CBOR<redeemer>>
```

Ryun1 commented on CIP-0141/README.md:
```diff
@@ -0,0 +1,216 @@
+
+### Acceptance Criteria
+
+- [ ] The interface is implemented and supported by various wallet providers.
```
```suggestion
- [ ] The interface is implemented and supported by three different wallet providers.
```

Ryun1 reviewed (approved):
happy to see this on merged for now

I feel this solves a legitimate problem that I suspect will be revisited in the future
so would be happy to see tweaks in the future
